library/TeslaAuthHelper.cs

```diff
@@ -167,137 +160,40 @@ public async Task<Tokens> RefreshTokenAsync(string refreshToken, CancellationTok
-                    }
-
-                    return await GetAuthorizationCodeWithMfaAsync(mfaCode, loginInfo, client, cancellationToken);
-                }
-                else if (result.StatusCode == HttpStatusCode.OK)
-                {
-                    // CAPTCHA requirement, probably.
-                    bool hasCaptcha = resultContent.Contains("captcha");
-                    bool hasReCaptcha = resultContent.Contains("recaptcha");
-                    bool hasRedirectLocation = result.Headers.Location != null;
-                    if (!hasCaptcha)
-                    {
-                        // This is unexpected.  What is this?
-                        throw new Exception("Can't log in - expected redirect did not occur.  hasRedirect: " + hasRedirectLocation);
-                    }
-                    if (hasReCaptcha)
-                    {
-                        throw new Exception("Can't log in - Tesla ReCAPTCHA may be necessary");
-                    }
-                    throw new Exception("Can't log in - Tesla CAPTCHA support needed.");
-                }
-                else
-                {
-                    // Possible causes for ending up here:
-                    // 1) Account could require an MFA code and we didn't provide it, maybe.  Captcha check may come first though.
-                    // What happens with the wrong MFA code?  Does that come here?
-                    // We believe cases like the account being locked fall into a different codepath below.
-                    throw new Exception("Expected redirect did not occur - probably need multi-factor authentication code.  Status code: " + result.StatusCode);
-                }
-            }
-
-            var location = result.Headers.Location;
-
-            if (location == null)
-            {
-                throw new Exception("Redirect location not available");
-            }
-
-            string code = HttpUtility.ParseQueryString(location.Query).Get("code");
-            return code;
-        }
+        
 
         async Task<Tokens> ExchangeCodeForBearerTokenAsync(string code, HttpClient client, CancellationToken cancellationToken)
         {
             var body = new JObject
             {
                 {"grant_type", "authorization_code"},
-                {"client_id", "ownerapi"},
+                {"client_id", clientId == TESLA_CLIENT_ID ? "ownerapi" : clientId},
+                {"client_secret", clientSecret },
                 {"code", code},
                 {"code_verifier", loginInfo.CodeVerifier},
-                {"redirect_uri", "https://auth.tesla.com/void/callback"},
-                //{"locale", "en-US" },
-            };
+                {"redirect_uri", redirectUri},
+                { "scope", scopes },
+                { "audience", "https://fleet-api.prd.na.vn.cloud.tesla.com" }

```

Comment:
Not sure, but isn't this url based on region? See https://developer.tesla.com/docs/fleet-api#overview

- North America, Asia-Pacific (excluding China): https://fleet-api.prd.na.vn.cloud.tesla.com/
- Europe, Middle East, Africa: https://fleet-api.prd.eu.vn.cloud.tesla.com/

Reply:
Ah good catch this was on my Todo list and I missed it. Although it would be good to test if this actually matters.